pym/bob/scm/git.py

```diff
@@ -266,8 +269,21 @@ async def __checkoutTag(self, invoker, fetchCmd, switch, oldSpec):
                     await invoker.checkCommand(["git", "symbolic-ref",
                         "-q", "HEAD"], cwd=self.__dir)
 
-            await invoker.checkCommand(["git", "checkout", "-q", "--no-recurse-submodules",
-                self.__commit if self.__commit else "tags/"+self.__tag], cwd=self.__dir)
+            # test if the configured commit is on the configured branch
+            if self.__branch and self.__commit and self.__useBranchAndCommit:
+                branches = await invoker.checkOutputCommand(["git", "branch", "-a",
+                    "--contains", self.__commit], cwd=self.__dir)
+                if not re.search( r".*/origin/"+re.escape(self.__branch), branches, re.MULTILINE):
+                    raise BuildError ("Commit: " + self.__commit + " is not on branch "
+                        + self.__branch)
+                # if we have both checkout the branch and move to the commit
+                await self.__checkoutBranch(invoker, fetchCmd, switch)
+
+                await invoker.checkCommand(["git", "reset", "--keep",
+                    "--recurse-submodules", self.__commit], cwd=self.__dir)
```

Comment:
This will always move the branch to the current head and then back to the configured commit. If sub-modules were added in between, this will leave behind a wrecked workspace because even git cannot cope with vanishing submodules. Instead I would suggest that `__checkoutBranch()` somehow needs to check out the configured commit directly.

Reply:
Thinking more about it I think this new behaviour should not be folded into the existing code paths but be an entire new method, e.g. `__checkoutTagOnBranch`. Because effectively its a mixture of `__checkoutTag` and `__checkoutBranch`. Then we don't have to change the existing behaviour...